test/testutil.c

```diff
@@ -145,18 +148,284 @@ int run_tests(const char *test_prog_name)
+DEFINE_COMPARISONS(unsigned char, uchar, "%u")
+DEFINE_COMPARISONS(long, long, "%ld")
+DEFINE_COMPARISONS(unsigned long, ulong, "%lu")
+DEFINE_COMPARISONS(size_t, size_t, "%" OSSLzu)
+
+DEFINE_COMPARISON(void *, ptr, eq, ==, "%p")
+DEFINE_COMPARISON(void *, ptr, ne, !=, "%p")
+
+int test_ptr_null(const char *file, int line, const char *s, const void *p)
+{
+    if (p == NULL)
+        return 1;
+    test_fail_message(1, file, line, "ptr", "%s [%p] == NULL", s, p);
+    return 0;
+}
+
+int test_ptr(const char *file, int line, const char *s, const void *p)
+{
+    if (p != NULL)
+        return 1;
+    test_fail_message(1, file, line, "ptr", "%s [%p] != NULL", s, p);
+    return 0;
+}
+
+int test_true(const char *file, int line, const char *s, int b)
+{
+    if (b)
+        return 1;
+    test_fail_message(1, file, line, "bool", "%s [false] == true", s);
+    return 0;
+}
+
+int test_false(const char *file, int line, const char *s, int b)
+{
+    if (!b)
+        return 1;
+    test_fail_message(1, file, line, "bool", "%s [true] == false", s);
+    return 0;
+}
+
 static const char *print_string_maybe_null(const char *s)
 {
     return s == NULL ? "(NULL)" : s;
 }
 
-int strings_equal(const char *desc, const char *s1, const char *s2)
+int test_str_eq(const char *file, int line, const char *st1, const char *st2,
+                const char *s1, const char *s2)
 {
     if (s1 == NULL && s2 == NULL)
       return 1;
     if (s1 == NULL || s2 == NULL || strcmp(s1, s2) != 0) {
-        fprintf(stderr, "%s mismatch: %s vs %s\n", desc, print_string_maybe_null(s1),
-                print_string_maybe_null(s2));
+        test_fail_message(1, file, line, "string", "%s [%s] == %s [%s]",
+                          st1, print_string_maybe_null(s1),
+                          st2, print_string_maybe_null(s2));
+        return 0;
+    }
+    return 1;
+}
+
+int test_str_ne(const char *file, int line, const char *st1, const char *st2,
+                const char *s1, const char *s2)
+{
+    if ((s1 == NULL) ^ (s2 == NULL))
+      return 1;
+    if (s1 == NULL || strcmp(s1, s2) == 0) {
+        test_fail_message(1, file, line, "string", "%s [%s] != %s [%s]",
+                          st1, print_string_maybe_null(s1),
+                          st2, print_string_maybe_null(s2));
+        return 0;
+    }
+    return 1;
+}
+
+/*
+ * We could use OPENSSL_buf2hexstr() to do this but trying to allocate memory
+ * in a failure state isn't generally a great idea.
+ */
+static const char *print_mem_maybe_null(const void *s, size_t n,
+                                        char out[MEM_BUFFER_SIZE])
+{
+    size_t i;
+    const unsigned char *p = (const unsigned char *)s;
+    int pad = 2*n >= MEM_BUFFER_SIZE;
+
+    if (s == NULL)
+        return "(NULL)";
+    if (pad)
+        n = MEM_BUFFER_SIZE-4;
+    
+    for (i=0; i<2*n; i++) {
+        unsigned char c = (i & 1) != 0 ? p[i / 2] & 15 : p[i / 2] >> 4;
+        out[i] = c["0123456789abcdef"];
```

Comment:
This originally comes from K&R C and has remained ever since.  `a[b]` is defined to be `*(a + b)`.  Thus, the order doesn't matter.

Little quirk in my coding to hex conversions...